RFC/0002-signed-envelopes.md

```diff
@@ -0,0 +1,108 @@
+
+## Domain Separation
+
+Signatures can be used for a variety of purposes, and a signature made for a
+specific purpose MUST NOT be considered valid for a different purpose.
+
+Without this property, an attacker could convince a peer to sign a payload in
+one context and present it as valid in another, for example, presenting a signed
+address record as a pubsub message.
+
+We separate signatures into "domains" by prefixing the data to be signed with a
+string unique to each domain. This string is not contained within the payload or
+the outer envelope structure. Instead, each libp2p subsystem that makes use of
+signed envelopes will provide their own domain string when constructing the
+envelope, and again when validating the envelope. If the domain string used to
+validate is different from the one used to sign, the signature validation will
+fail.
+
+Domain strings may be any valid UTF-8 string, but should be fairly short and
+descriptive of their use case, for example `"libp2p-routing-record"`.
+
+## Type Hinting
+
+The envelope record can contain an arbitrary byte string payload, which will
+need to be interpreted in the context of a specific use case. To assist in
+"hydrating" the payload into an appropriate domain object, we include a "type
+hint" field. The type hint consists of a [multicodec][multicodec] code,
+optionally followed by an arbitrary byte sequence.
+
+This allows very compact type hints that contain just a multicodec, as well as
+"path" multicodecs of the form `/some/thing`, using the ["namespace"
+multicodec](https://github.com/multiformats/multicodec/blob/master/table.csv#L23),
+whose binary value is equivalent to the UTF-8 `/` character.
+
+## Wire Format
+
+Since we already have a [protobuf definition for public keys][peer-id-spec], we
+can use protobuf for this as well and easily embed the key in the envelope:
+
+
+```protobuf
+message SignedEnvelope {
+  PublicKey publicKey = 1; // see peer id spec for definition
+  bytes typeHint = 2;      // type hint
+  bytes contents = 3;      // payload
+  bytes signature = 4;     // see below for signing rules
+}
+```
+
+The `publicKey` field contains the public key whose secret counterpart was used
+to sign the message. This MUST be consistent with the peer id of the signing
+peer, as the recipient will derive the peer id of the signer from this key.
+
+The `typeHint` field contains a [multicodec][multicodec]-prefixed type hint as
+described in the [Type Hinting section](#type-hinting).
+
+The `contents` field contains the arbitrary byte string payload.
+
+The `signature` field contains a signature of all fields except `publicKey`,
+generated as described below.
+
+## Signature Production / Verification
+
+When signing, a peer will prepare a buffer by concatenating the following:
+
+- The length of the [domain separation string](#domain-separation) string in
+  bytes
+- The domain separation string, encoded as UTF-8
+- The length of the `typeHint` field in bytes
```

Comment:
I'd feel better if we _also_ signed the key. I'm concerned users will use this in a "commitment" and/or "ticket" protocol as follows:

1. Victim publishes a signed envelop stating "I (unspecified) claim X" to a blockchain.
2. Attacker generates a key such that the key validates the signed envelope.
3. Attacker claims X.

Basically, "an attacker can't make a public key that validates a target signature S" is _not_ a guarantee of public key cryptography. However, it's one that's often assumed.

Reply:
You know, this is probably more work than necessary.